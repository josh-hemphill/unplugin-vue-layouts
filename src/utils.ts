import { posix } from 'node:path';

export function normalizePath(path: string) {
  path = path.startsWith('/') ? path : `/${path}`;
  return posix.normalize(path);
}
