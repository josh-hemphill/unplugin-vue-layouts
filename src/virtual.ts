import { normalizePath } from './utils';
import type { Options } from './index';

export function createVirtualModuleID(name: string) {
  const virtualModuleId = `virtual:${name}`;
  const resolvedVirtualModuleId = `\0${virtualModuleId}`;
  return {
    virtualModuleId,
    resolvedVirtualModuleId,
  };
}

export async function createVirtualGlob(target: string, isSync: boolean) {
  return `import.meta.glob("${target}/**/*.vue", { eager: ${isSync} })`;
}

type VirtualModuleCodeOptions = Required<Options>;

export async function createVirtualModuleCode(options: VirtualModuleCodeOptions) {
  const { target, defaultLayout, importMode, skipTopLevelRouteLayout, wrapComponent } = options;
  const normalizedTarget = normalizePath(target);
  const isSync = importMode === 'sync';

  const skipCode = /* ts */ `
    // unplugin-vue-router adds a top-level route to the routing group, which we should skip (ref â†’ https://github.com/JohnCampionJr/vite-plugin-vue-layouts/issues/134)
    const skipLayout = top && !route.component && route.children?.find(r => (r.path === '' || r.path === '/') && r.meta?.isLayout)
    if (skipLayout) {
      return route
    }
  `;

  function createLayoutRoute() {
    return /* ts */ `
    // Create a new route with the layout as the component
    const layoutRoute = {
      path: route.path,
      component: layouts[layout],
      meta: { ...route.meta, isLayout: true },
      // Handle root path specially to avoid infinite nesting
      children: top && route.path === '/' 
        ? [route] 
        : [{ ...route, path: '', meta: { ...route.meta, isLayout: false } }]
    }
    return layoutRoute
    `;
  }

  function createLayoutComponent() {
    return /* ts */ `
    if (!route.component) {
      return route
    }
    return {
      ...route,
      component: h('div', [
        h(layouts[layout]),
        h(define${isSync ? '' : 'Async'}Component(() => route.component())),
      ]),
      meta: {
        ...route.meta,
        isLayout: true
      }
    }
    `;
  }

  function fixIndentation(code: string, indent: number) {
    return code.replace(/\n/g, `\n${' '.repeat(indent)}`);
  }

  const importStatements = [
    wrapComponent ? `import { h, define${isSync ? '' : 'Async'}Component } from 'vue'` : '',
    `import type { RouteRecordRaw, Router } from 'vue-router'`,
  ]
    .filter(Boolean)
    .join('\n');

  return /* ts */ `
    ${importStatements}
    export function createGetRoutes(router: Router, withLayout = false) {
      const routes = router.getRoutes()
      if (withLayout) {
        return routes
      }
      return () => routes.filter(route => !route.meta?.isLayout)
    }

    export function setupLayouts(routes: RouteRecordRaw[]) {
      const layouts = {}

      const modules = ${await createVirtualGlob(normalizedTarget, isSync)}

      Object.entries(modules).forEach(([name, module]) => {
        let key = name.replace("${normalizedTarget}/", '').replace('.vue', '')
        layouts[key] = ${isSync ? 'module.default' : 'module'}
      })

      function deepSetupLayout(lRoutes: RouteRecordRaw[], top = true) {
        return lRoutes.map(route => {
          // Process children first
          if (route.children?.length > 0) {
            route.children = deepSetupLayout(route.children, false)
          }

          // Skip if already processed
          if (route.meta?.isLayout) {
            return route
          }

          const layout = route.meta?.layout ?? '${defaultLayout}'

          ${fixIndentation(skipTopLevelRouteLayout ? skipCode : '', 6)}
          
          if (layout && layouts[layout]) {
            ${fixIndentation(wrapComponent ? createLayoutComponent() : createLayoutRoute(), 8)}
          }

          return route
        })
      }

      return deepSetupLayout(routes)
    }
  `;
}
