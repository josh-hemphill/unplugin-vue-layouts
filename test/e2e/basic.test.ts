import path from 'node:path';
import { beforeAll, describe, expect, it, vi } from 'vitest';
import type { RouteRecordRaw } from 'vue-router';
import { createVirtualModuleCode, createVirtualModuleID } from '../../src/virtual.js';
import { baseDir, codeToModule, getRoutes, sanitizeRouteComponentPaths } from './utils';

describe('basic', () => {
  let routesContext: RouteRecordRaw[];

  beforeAll(async () => {
    const { routes } = await getRoutes();
    routesContext = routes;
  });

  it('should work', async () => {
    const code = await createVirtualModuleCode({
      target: path.join('fixtures', 'layouts'),
      defaultLayout: 'default',
      importMode: 'sync',
      wrapComponent: false,
      skipTopLevelRouteLayout: false,
    });

    // First verify the code is valid JavaScript
    let Layout;
    try {
      Layout = (await codeToModule(code, 'layout')) as typeof import('virtual:vue-layouts');
    } catch (error) {
      console.error('Code evaluation failed:');
      console.error(error);
      throw error;
    }

    expect(Layout).toBeDefined();
    // Rest of the test...
    const routes = Layout.setupLayouts(routesContext);
    expect(sanitizeRouteComponentPaths(routes)).toMatchSnapshot('Unwrapped routes');
    Layout.setupLayouts(routes);
  });

  const getWrappedRoutes = async (code: PromiseLike<string>) => {
    const Layout = (await codeToModule(await code, 'layout')) as typeof import('virtual:vue-layouts');
    const routes = Layout.setupLayouts(routesContext);
    const routesWrapped = Layout.setupLayouts(routes);
    return { routes, routesWrapped };
  };

  it('should work with async layouts', async () => {
    await getWrappedRoutes(
      createVirtualModuleCode({
        target: path.join('fixtures', 'layouts'),
        defaultLayout: 'default',
        importMode: 'async',
        wrapComponent: false,
        skipTopLevelRouteLayout: false,
      }),
    );
  });

  it('should work with async layouts and wrapComponent', async () => {
    await getWrappedRoutes(
      createVirtualModuleCode({
        target: path.join('fixtures', 'layouts'),
        defaultLayout: 'default',
        importMode: 'async',
        wrapComponent: true,
        skipTopLevelRouteLayout: false,
      }),
    );
  });

  it('should work with async layouts and skipTopLevelRouteLayout', async () => {
    await getWrappedRoutes(
      createVirtualModuleCode({
        target: path.join('fixtures', 'layouts'),
        defaultLayout: 'default',
        importMode: 'async',
        wrapComponent: false,
        skipTopLevelRouteLayout: true,
      }),
    );
  });

  it('should match snapshot with wrapComponent', async () => {
    const { routesWrapped } = await getWrappedRoutes(
      createVirtualModuleCode({
        target: path.join('fixtures', 'layouts'),
        defaultLayout: 'default',
        importMode: 'sync',
        wrapComponent: true,
        skipTopLevelRouteLayout: false,
      }),
    );
    expect(sanitizeRouteComponentPaths(routesWrapped)).toMatchSnapshot('Wrapped Component');
  });

  it('should match snapshot with route wrapping', async () => {
    const { routesWrapped } = await getWrappedRoutes(
      createVirtualModuleCode({
        target: path.join('fixtures', 'layouts'),
        defaultLayout: 'default',
        importMode: 'sync',
        wrapComponent: false,
        skipTopLevelRouteLayout: false,
      }),
    );
    expect(sanitizeRouteComponentPaths(routesWrapped)).toMatchSnapshot('Route Wrapping');
  });
});
