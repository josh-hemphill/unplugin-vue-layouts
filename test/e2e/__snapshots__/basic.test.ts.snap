// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic > should match snapshot with route wrapping > Route Wrapping 1`] = `
[
  {
    "component": {
      "__file": "./fixtures/pages/index.vue",
      "render": [Function],
    },
    "name": "/",
    "path": "/",
  },
  {
    "component": {
      "__file": "./fixtures/pages/about.vue",
      "render": [Function],
    },
    "name": "/about",
    "path": "/about",
  },
  {
    "children": [
      {
        "component": {
          "__file": "./fixtures/pages/meta_tagged/route_block.vue",
          "render": [Function],
        },
        "meta": {
          "layout": "custom",
        },
        "name": "/meta_tagged/route_block",
        "path": "route_block",
      },
    ],
    "path": "/meta_tagged",
  },
  {
    "children": [
      {
        "component": {
          "__file": "./fixtures/pages/nested/index.vue",
          "render": [Function],
        },
        "name": "/nested/",
        "path": "",
      },
      {
        "component": {
          "__file": "./fixtures/pages/nested/foo.vue",
          "render": [Function],
        },
        "name": "/nested/foo",
        "path": "foo",
      },
    ],
    "path": "/nested",
  },
  {
    "children": [
      {
        "component": {
          "__file": "./fixtures/pages/parent/child.vue",
          "render": [Function],
        },
        "name": "/parent/child",
        "path": "child",
      },
    ],
    "component": {
      "__file": "./fixtures/pages/parent.vue",
      "render": [Function],
    },
    "name": "/parent",
    "path": "/parent",
  },
]
`;

exports[`basic > should match snapshot with wrapComponent > Wrapped Component 1`] = `
[
  {
    "component": {
      "__file": "./fixtures/pages/index.vue",
      "render": [Function],
    },
    "name": "/",
    "path": "/",
  },
  {
    "component": {
      "__file": "./fixtures/pages/about.vue",
      "render": [Function],
    },
    "name": "/about",
    "path": "/about",
  },
  {
    "children": [
      {
        "component": {
          "__file": "./fixtures/pages/meta_tagged/route_block.vue",
          "render": [Function],
        },
        "meta": {
          "layout": "custom",
        },
        "name": "/meta_tagged/route_block",
        "path": "route_block",
      },
    ],
    "path": "/meta_tagged",
  },
  {
    "children": [
      {
        "component": {
          "__file": "./fixtures/pages/nested/index.vue",
          "render": [Function],
        },
        "name": "/nested/",
        "path": "",
      },
      {
        "component": {
          "__file": "./fixtures/pages/nested/foo.vue",
          "render": [Function],
        },
        "name": "/nested/foo",
        "path": "foo",
      },
    ],
    "path": "/nested",
  },
  {
    "children": [
      {
        "component": {
          "__file": "./fixtures/pages/parent/child.vue",
          "render": [Function],
        },
        "name": "/parent/child",
        "path": "child",
      },
    ],
    "component": {
      "__file": "./fixtures/pages/parent.vue",
      "render": [Function],
    },
    "name": "/parent",
    "path": "/parent",
  },
]
`;

exports[`basic > should work > Unwrapped routes 1`] = `
[
  {
    "component": {
      "__file": "./fixtures/pages/index.vue",
      "render": [Function],
    },
    "name": "/",
    "path": "/",
  },
  {
    "component": {
      "__file": "./fixtures/pages/about.vue",
      "render": [Function],
    },
    "name": "/about",
    "path": "/about",
  },
  {
    "children": [
      {
        "component": {
          "__file": "./fixtures/pages/meta_tagged/route_block.vue",
          "render": [Function],
        },
        "meta": {
          "layout": "custom",
        },
        "name": "/meta_tagged/route_block",
        "path": "route_block",
      },
    ],
    "path": "/meta_tagged",
  },
  {
    "children": [
      {
        "component": {
          "__file": "./fixtures/pages/nested/index.vue",
          "render": [Function],
        },
        "name": "/nested/",
        "path": "",
      },
      {
        "component": {
          "__file": "./fixtures/pages/nested/foo.vue",
          "render": [Function],
        },
        "name": "/nested/foo",
        "path": "foo",
      },
    ],
    "path": "/nested",
  },
  {
    "children": [
      {
        "component": {
          "__file": "./fixtures/pages/parent/child.vue",
          "render": [Function],
        },
        "name": "/parent/child",
        "path": "child",
      },
    ],
    "component": {
      "__file": "./fixtures/pages/parent.vue",
      "render": [Function],
    },
    "name": "/parent",
    "path": "/parent",
  },
]
`;
